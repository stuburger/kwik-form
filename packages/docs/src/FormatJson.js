/* eslint-disable react/prop-types */
import React, { Component } from 'react'
import isEqual from 'react-fast-compare'
import classnames from 'classnames'

export default class FormatJson extends Component {
  constructor(props) {
    super(props)
    this.state = { showChangeEffect: false }
  }

  componentDidMount() {
    this.timeout = null
  }

  shouldComponentUpdate(np, ns) {
    const { value } = this.props
    const { showChangeEffect } = this.state
    return ns.showChangeEffect !== showChangeEffect || !isEqual(value, np.value)
  }

  componentDidUpdate(prev) {
    const { value } = this.props

    if (!isEqual(value, prev.value)) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState(() => {
        return { showChangeEffect: true }
      })
      this.timeout = setTimeout(() => {
        if (this.isUnmounting) return
        this.setState(() => {
          return { showChangeEffect: false }
        })
      }, 500)
    }
  }

  componentWillUnmount() {
    this.isUnmounting = true
    clearTimeout(this.timeout)
  }

  render() {
    const { value, spacing = 2 } = this.props
    const { showChangeEffect } = this.state
    return (
      <pre
        className={classnames('json-formatted', {
          'change-fade': !showChangeEffect,
        })}
      >
        {JSON.stringify(value, null, spacing)}
      </pre>
    )
  }
}
