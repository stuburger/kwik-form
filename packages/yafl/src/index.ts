import * as React from 'react'
import { FormProvider } from './sharedTypes'
import createForm from './createForm'
import createSection from './createSection'
import createField from './createField'
import createRepeat from './createRepeat'
import createFormError from './createFormError'
import createForwardProps from './createForwardProps'
import createHooks from './createHooks'
import { BLOCKER } from './useSafeContext'

export function createFormContext<F extends object>() {
  const context = React.createContext<FormProvider<F> | Symbol>(BLOCKER)

  const { Provider } = context

  const hooks = createHooks<F>(context)
  return {
    ...hooks,
    Form: createForm<F>(Provider as React.Provider<FormProvider<F, F>>),
    Section: createSection<F>(context),
    Repeat: createRepeat<F>(context),
    Field: createField<F>(context),
    FormError: createFormError(context),
    ForwardProps: createForwardProps(context as any),
  }
}

export const {
  Field,
  Form,
  ForwardProps,
  Repeat,
  Section,
  FormError,
  useYaflContext,
} = createFormContext<any>()

export * from './sharedTypes'
