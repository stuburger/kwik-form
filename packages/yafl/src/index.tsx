/* eslint-disable react/prop-types */
import * as React from 'react'
import { FormProvider, Name, YaflBaseContext } from './sharedTypes'
import createForm from './Form'
import createSection from './Section'
import createField from './Field'
import createRepeat from './Repeat'
import createFormError from './FormError'
import createUseField from './useField'
import { BLOCKER, useSafeContext } from './useSafeContext'
import { useBranch } from './utils'
import createUseForm from './useForm'

export function createFormContext<F extends object>() {
  const context = React.createContext<FormProvider<F> | Symbol>(BLOCKER)

  const { Provider } = context

  const { useField } = createUseField<F>(context)
  const useForm = createUseForm()

  function useCommonValues<T = any>(): T {
    const { commonValues } = useSafeContext<any, any>(context)
    return commonValues as any
  }

  function useBranchValues<T extends object = any>(name: Name): T {
    const { branchValues } = useBranch<any, any>(name, context)
    return branchValues
  }

  function useYaflContext() {
    return useSafeContext(context)
  }

  const YaflProvider: React.FC<{
    value: YaflBaseContext<F> & {
      branchValues?: Record<string, any>
      commonValues?: Record<string, any>
    }
  }> = (props) => {
    const { children, value } = props
    const { branchValues = {}, commonValues = {} } = value

    return (
      <Provider value={{ ...value, path: '', branchValues, commonValues }}>{children}</Provider>
    )
  }

  return {
    useField,
    useForm,
    useCommonValues,
    useBranchValues,
    useYaflContext,
    YaflProvider,
    Form: createForm<F>(useForm, Provider),
    Section: createSection<F>(context),
    Repeat: createRepeat<F>(context),
    Field: createField<F>(useField, useCommonValues, useBranchValues),
    FormError: createFormError(context),
  }
}

export const {
  Field,
  Form,
  useField,
  Repeat,
  useForm,
  Section,
  FormError,
  YaflProvider,
  useYaflContext,
  useCommonValues,
  useBranchValues,
} = createFormContext<any>()

export * from './sharedTypes'
