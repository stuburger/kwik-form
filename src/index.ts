import * as React from 'react'
import { FormProvider } from './sharedTypes'
import { getDefaultProviderValue } from './defaults'
import createForm from './createForm'
import createSection from './createSection'
import createField from './createField'
import createConnect from './createConnect'
import createRepeat from './createRepeat'
import createValidator from './createValidator'

export function createFormContext<F extends object>() {
  const context = React.createContext<FormProvider<F>>(getDefaultProviderValue())
  const { Provider, Consumer } = context
  return {
    Form: createForm<F>(Provider),
    Section: createSection<F>(Provider, Consumer),
    Repeat: createRepeat<F>(Provider, Consumer),
    Field: createField<F>(Provider, Consumer),
    Validator: createValidator(Consumer),
    connect: createConnect<F>(Consumer)
  }
}

export const { Form, Section, Field, Repeat, Validator, connect } = createFormContext<any>()
export * from './sharedTypes'
