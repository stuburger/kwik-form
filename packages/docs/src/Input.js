/* eslint-disable react/prop-types */
import React from 'react'
import classnames from 'classnames'
import { useField } from 'yafl'

export default function TextInput(props) {
  const { name, busy, label, type, placeholder, validate } = props
  const [input, meta] = useField(name, { validate })
  const { isValid, visited, submitCount, errors = [] } = meta
  const showError = !isValid && (visited || submitCount > 0)

  return (
    <div className="input-group">
      <label htmlFor={input.name}>{label}</label>
      {busy && (
        <span className="input-validating" role="img" aria-label="Validating">
          ðŸ¤ž
        </span>
      )}
      <input
        {...input}
        type={type}
        placeholder={placeholder}
        className={classnames('text-input', {
          error: showError,
        })}
      />
      {showError && <div className="input-feedback">{errors[0]}</div>}
    </div>
  )
}
