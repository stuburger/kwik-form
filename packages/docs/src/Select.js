/* eslint-disable react/prop-types */
import React from 'react'
import { Validator } from 'yafl'
import classnames from 'classnames'

export default function Select(props) {
  const { label, input, meta, placeholder, type, busy, options = [], validators = [] } = props
  const { isValid, visited, submitCount, errors = [] } = meta
  const showError = !isValid && (visited || submitCount > 0)
  return (
    <div className="input-group">
      <label htmlFor={input.name}>{label}</label>
      {busy && (
        <span className="input-validating" role="img" aria-label="Validating">
          ðŸ¤ž
        </span>
      )}
      <select
        {...input}
        type={type}
        className={classnames('text-input', {
          error: showError,
        })}
      >
        <option value="">{placeholder}</option>
        {options.map((value) => (
          <option key={value} value={value}>
            {value}
          </option>
        ))}
      </select>
      {showError && <div className="input-feedback">{errors[0]}</div>}
      {validators.reduceRight((ret, validate) => {
        return <Validator msg={validate(input.value)}>{ret}</Validator>
      }, null)}
    </div>
  )
}
