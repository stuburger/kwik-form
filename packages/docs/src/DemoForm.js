/* eslint-disable react/sort-comp */
/* eslint-disable no-nested-ternary */
/* eslint-disable react/prop-types */
import React, { useState } from 'react'
import { YaflProvider, Section, FormError, useYaflContext, useForm } from 'yafl'
import Input from './Input'
import { required, minLength, email, min } from './validators'
import FormatJson from './FormatJson'

const initialValue = {
  fullName: '',
  age: 0,
  contact: {
    tel: '',
    email: '',
    random: {
      address: {
        streetNo: '',
      },
    },
  },
}

function DemoForm() {
  const [isSubmitting, setSubmitting] = useState(false)
  const [formData, setFormData] = useState(initialValue)

  const yafl = useForm({
    initialValue: formData,
    onSubmit: (value) => {
      setSubmitting(true)
      setTimeout(() => {
        setFormData(value)
        setSubmitting(false)
      }, 1200)
    },
  })

  return (
    <YaflProvider value={yafl}>
      <form onSubmit={yafl.submit}>
        <div className="row">
          <div className="col col--4">
            <p className="hero__subtitle">Form Value</p>
            <FormatJson value={yafl.formValue} />
          </div>
          <div className="col col--4">
            <p className="hero__subtitle">Demo Form</p>
            <Input
              name="fullName"
              label="Full Name"
              placeholder="E.g. John Smith"
              validate={[required, minLength(3)]}
            />
            <Input
              name="age"
              type="number"
              label="Age"
              validate={[required, min(5, 'Must be at least 5 years old!')]}
              placeholder="E.g. 29"
            />
            <Section name="contact">
              <Input
                name="email"
                label="Email"
                placeholder="E.g. example@email.com"
                validate={[required, email()]}
                // busy={asyncValidating}
              />
              <FormError
                path="contact.email"
                // msg={asyncEmailError}
              />
              <Section name="random.address">
                <Input name="postalCode" label="Postal Code" placeholder="Postal Code" />
              </Section>
            </Section>
            <Submit submitting={isSubmitting} fetching={false} />
          </div>

          <div className="col col--4">
            <p className="hero__subtitle">Meta State</p>
            <FormatJson
              value={{
                activeField: yafl.activeField,
                submitCount: yafl.submitCount,
                errorCount: yafl.errorCount,
                formIsValid: yafl.formIsValid,
                formIsDirty: yafl.formIsDirty,
                touched: yafl.touched,
              }}
            />
          </div>
          <div className="col col--12">
            <p className="hero__subtitle">Errors</p>
            <FormatJson value={{ errorCount: yafl.errorCount, errors: yafl.errors }} />
          </div>
        </div>
      </form>
    </YaflProvider>
  )
}

export default DemoForm

const Submit = ({ fetching, submitting }) => {
  const yafl = useYaflContext()
  return (
    <button
      type="submit"
      className="button button--outline button--primary text--right"
      disabled={fetching || submitting}
    >
      {fetching
        ? 'Loading ⏳'
        : submitting
        ? 'Saving ⏳'
        : yafl.formIsDirty
        ? 'Save 🚀'
        : 'Saved 👍'}
    </button>
  )
}
