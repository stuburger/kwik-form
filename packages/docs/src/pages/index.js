import React from 'react'
import classnames from 'classnames'
import Layout from '@theme/Layout'
import Link from '@docusaurus/Link'
import useDocusaurusContext from '@docusaurus/useDocusaurusContext'
import useBaseUrl from '@docusaurus/useBaseUrl'
import styles from './styles.module.css'
import DemoForm from '../DemoForm'

function Home() {
  const context = useDocusaurusContext()
  const { siteConfig = {} } = context
  return (
    <Layout
      title={`Hello from ${siteConfig.title}`}
      description="Description will go into a meta tag in <head />"
    >
      <header className={classnames('hero hero--primary', styles.heroBanner)}>
        <div className="container">
          <h1 className="hero__title">{siteConfig.title}</h1>
          <p className="hero__subtitle">{siteConfig.tagline}</p>

          <div className={styles.buttons}>
            <Link
              className={classnames('button button--secondary button--lg', styles.getStarted)}
              to={useBaseUrl('docs/')}
            >
              Get Started
            </Link>
          </div>
        </div>
      </header>
      <div className="container margin-top--md">
        <DemoForm />
      </div>
      {/* <main>
      </main> */}
    </Layout>
  )
}

export default Home
