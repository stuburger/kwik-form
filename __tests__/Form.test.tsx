import * as React from 'react'
import { cleanup } from 'react-testing-library'
import { Person, Hobby } from '../src'
import renderer from 'react-test-renderer'
import { createFormRenderer } from './helpers'

function TextInput(props: any) {
  return (
    <div>
      <label>{props.label}</label>
      <input {...props.input} />
    </div>
  )
}
const personData: Person = {
  name: 'Stuart Bourhill',
  age: undefined!,
  contact: {
    tel: '07934449898',
    address: {
      code: 'SW9 1EF',
      street: 'adsfa'
    }
  },
  hobbies: [{ type: 'art', name: 'print making' }]
}

const defaultPerson: Person = {
  name: '',
  age: 0,
  contact: {
    tel: '',
    address: {
      code: '',
      street: ''
    }
  },
  hobbies: []
}

describe('<Form />', () => {
  afterEach(cleanup)

  describe('snapshot', () => {
    const { Form, Section, Field, Repeat } = createFormRenderer<Person>()
    it('should match previous snapshot', () => {
      const tree = renderer
        .create(
          <Form initialValue={personData} defaultValue={defaultPerson} onSubmit={() => {}}>
            {yafl => (
              <form onSubmit={yafl.submit}>
                <Field name="firstName" label="First Name" component={TextInput} />
                <Section name="contact">
                  <Field name="tel" label="First Name" component={TextInput} />
                  <Field name="email" label="Email Address" component={TextInput} />
                </Section>
                <Repeat<Hobby>
                  name="hobbies"
                  fallback={[{ type: 'art', name: 'print making' }, undefined, undefined] as any}
                >
                  {hobbies => {
                    return hobbies.map((value, i) => {
                      return (
                        <Section name={i} key={i} fallback={{ type: 'new', name: 'new' }}>
                          <Field name="type" label="Type of hobby" component={TextInput} />
                          <Field name="name" label="Name of hobby" component={TextInput} />
                        </Section>
                      )
                    })
                  }}
                </Repeat>
                <pre>{JSON.stringify(yafl, null, 2)}</pre>
              </form>
            )}
          </Form>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('FormProps', () => {
    type B = { x?: any[]; y?: Date }
    type Thing = { a?: number; b?: B }

    const initialValue: Thing = {
      a: undefined,
      b: {
        x: []
      }
    }

    const date = new Date()
    const defaultValue: Thing = {
      a: 12,
      b: {
        x: [{ name: 'Munk Jones' }],
        y: date
      }
    }

    const { renderForm, Field } = createFormRenderer<Thing>()
    const submitMk = jest.fn()
    const onSubmit = (formValue: Thing) => {
      submitMk(formValue)
    }
    const { getFormProps } = renderForm({
      initialValue,
      defaultValue,
      onSubmit,
      submitUnregisteredValues: true
    })

    const props = getFormProps()

    const expectedStartingValue = {
      a: 12,
      b: { x: [{ name: 'Munk Jones' }], y: date }
    }

    describe('render prop initial values', () => {
      it('supplies the correct values for the forms render props', () => {
        expect(props.initialValue).toEqual(initialValue)
        expect(props.defaultValue).toEqual(defaultValue)
        expect(props.errors).toEqual({})
        expect(props.errorCount).toEqual(0)
        expect(props.formIsDirty).toEqual(false)
        expect(props.initialMount).toEqual(true)
        expect(props.submitCount).toEqual(0)
        expect(props.touched).toEqual({})
        expect(props.visited).toEqual({})
      })

      it('merges defaultValue into initalValue to create the form starting value', () => {
        expect(props.formValue).toEqual(expectedStartingValue)
      })
    })

    describe('FormProps.submit', () => {
      props.submit()

      it('calls the submit function', () => {
        expect(submitMk).toBeCalledTimes(1)
      })

      describe('when submitUnregisteredValues is false', () => {
        it('should call submit with the only registered fields', () => {
          const submitMk = jest.fn()
          const onSubmit = (formValue: Thing) => {
            submitMk(formValue)
            return false
          }
          const { getFormProps } = renderForm(
            {
              initialValue,
              defaultValue,
              onSubmit
            },
            <Field name="a" component="input" />
          )

          const { submit } = getFormProps()
          submit()
          expect(submitMk).toHaveBeenCalledWith({ a: 12 })
        })
      })

      describe('when submitUnregisteredValues is true', () => {
        it('should call submit with the whole formValue', () => {
          expect(submitMk).toHaveBeenCalledWith(expectedStartingValue)
        })
      })

      describe('when onSubmit is called and returns nothing', () => {
        it('should update submitCount', () => {
          const { submitCount } = getFormProps()
          expect(submitCount).toEqual(1)
        })
      })

      describe('when onSubmit is called and returns false', () => {
        it('should not update submitCount', () => {
          const submitMk = jest.fn()
          const onSubmit = (formValue: Thing) => {
            submitMk(formValue)
            return false
          }
          const { getFormProps } = renderForm({
            initialValue,
            defaultValue,
            onSubmit
          })

          const { submit } = getFormProps()

          submit()

          const { submitCount } = getFormProps()

          expect(submitMk).toHaveBeenCalledTimes(1)
          expect(submitCount).toEqual(0)
        })
      })
    })
  })
})

// let temp: FieldProps<Person> = undefined!;
// <Field
//   name="firstName"
//   render={props => {
//     temp = props
//     return <TextInput {...props} />
//   }}
// />

// expect(props.formValue).toEqual(temp.meta.formValue)
