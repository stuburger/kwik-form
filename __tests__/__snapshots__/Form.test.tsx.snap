// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <label>
      First Name
    </label>
    <input
      name="firstName"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="John"
    />
  </div>
  <div>
    <label>
      First Name
    </label>
    <input
      name="tel"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="983838333"
    />
  </div>
  <div>
    <label>
      Email Address
    </label>
    <input
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="asdf@testing.com"
    />
  </div>
  <div>
    <label>
      Type of hobby
    </label>
    <input
      name="type"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="art"
    />
  </div>
  <div>
    <label>
      Name of hobby
    </label>
    <input
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="print making"
    />
  </div>
  <div>
    <label>
      Type of hobby
    </label>
    <input
      name="type"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="new"
    />
  </div>
  <div>
    <label>
      Name of hobby
    </label>
    <input
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="new"
    />
  </div>
  <div>
    <label>
      Type of hobby
    </label>
    <input
      name="type"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="new"
    />
  </div>
  <div>
    <label>
      Name of hobby
    </label>
    <input
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value="new"
    />
  </div>
  <pre>
    {
  "formValue": {
    "firstName": "John",
    "lastName": "Woodcock",
    "age": 30,
    "contact": {
      "tel": "983838333",
      "email": "asdf@testing.com"
    },
    "friends": [
      "stuart",
      "oumar"
    ]
  },
  "errorCount": 0,
  "defaultValue": {
    "firstName": "",
    "lastName": "",
    "age": 0,
    "contact": {
      "tel": "",
      "email": "asdf@testing.com"
    }
  },
  "formIsDirty": true,
  "formIsValid": true,
  "initialValue": {
    "firstName": "John",
    "lastName": "Woodcock",
    "age": 30,
    "contact": {
      "tel": "983838333"
    },
    "friends": [
      "stuart",
      "oumar"
    ]
  },
  "initialMount": true,
  "errors": {},
  "touched": {},
  "visited": {},
  "activeField": null,
  "submitCount": 0,
  "components": {},
  "branchProps": {},
  "sharedProps": {},
  "path": [],
  "value": {
    "firstName": "John",
    "lastName": "Woodcock",
    "age": 30,
    "contact": {
      "tel": "983838333",
      "email": "asdf@testing.com"
    },
    "friends": [
      "stuart",
      "oumar"
    ]
  }
}
  </pre>
</form>
`;
